### ISSUE TITLE
Some plugins fail to compile in Linux Alpine Docker image

### ISSUE THREAD

**Issue by manios** [manios/docker-nagios#3]
See more in: [nagiosbuild.log](https://github.com/manios/docker-nagios/files/2953608/nagiosbuild.log)

Try to use the following Alpine packages as dependencies in order to build Nagios plugins:

```bash
apk add libldap \
openldap-dev \
freeradius-client-dev \
mysql \
mysql-client \
perl-dbd-mysql \
postgresql-libs \
postgresql-client \
postgresql-dev \
perl-test-postgresql \
libdbi-dev \
libssh2-dev \
libsmbclient \
net-snmp-dev \
fping \
net-snmp-perl
```

**Comment by qeepcologne** [comment]
for most platforms, compiled packages are available in alpine repo:
https://pkgs.alpinelinux.org/packages?name=nagios-plugins-*&branch=edge

as a workaround (because i need mysql plugin), i just installed and linked it:
RUN apk add nagios-plugins-mysql
RUN ln -s /usr/lib/nagios/plugins/check_mysql_query /opt/nagios/libexec/check_mysql_query

**Comment by manios** [comment]
Hi @qeepcologne !

Thank you for your valuable help! I have fixed the issue and you can download the latest image which includes all plugins including those (`ldap`,`snmp`,`check_pgsql`,`dbi`,`radius`,`mysql`) whose build was failing.

Can you please give it a try?

```bash
docker pull manios/nagios:latest
# or
docker pull manios/nagios:4.4.6
```

I am closing the issue. If you find any discrepancy please post a comment here and we will take a look.

Thanks,
Christos

**Comment by qeepcologne** [comment]
Error loading shared library libmariadb.so.3: No such file or directory (needed by ./check_mysql_query)
missing package is: mariadb-connector-c

**Comment by qeepcologne** [comment]
i found a workaround:
latest mariadb-connector-c from edge is working:
RUN apk add mariadb-connector-c --repository=http://dl-cdn.alpinelinux.org/alpine/edge/main

**Comment by qeepcologne** [comment]
...

**Comment by qeepcologne** [comment]
problem is from upgrade to alpine 3.13
i could bypass it by downgrading musl:
apk add musl=1.1.24-r10 --repository=http://dl-cdn.alpinelinux.org/alpine/v3.12/main

**Comment by manios** [comment]
Hi @qeepcologne !

Thank you for the debugging. I have realised that there are missing runtime dependencies and I am preparing a solution for that. I will let you know once we have a solution.

Thanks,
Christos

**Comment by manios** [comment]
Hi @qeepcologne !

I think that I have fixed all missing dependencies for all plugins. Can you please give it a try

```bash
docker pull manios/nagios:latest
# or
docker pull manios/nagios:4.4.6
```

and provide me with feedback so we can close the issue?

Thanks,
Christos

**Comment by qeepcologne** [comment]
Segmentation fault in check_mysql_query still remains. That is also the case when installing the plugin from alpine via apk.
Possible workaround is using alpine 3.12 or downgrading to musl 1.1 (see above).

**Comment by manios** [comment]
Hi @qeepcologne !

Can you please provide me with the steps you are doing in order to reproduce the issue?

I am executing the plugin using the `latest` tag:

```bash
root@192.168.0.28 ~
$ docker run --name nagios -p 8080:80 -d manios/nagios
Unable to find image 'manios/nagios:latest' locally
latest: Pulling from manios/nagios
[...] some lines omitted
71fa2e9c7e74: Pull complete 
Digest: sha256:1c88227d8d402a0c6f9131ddd1950833f3a24f8ff9333f39adde9ca2bc5ca674
Status: Downloaded newer image for manios/nagios:latest
f743fc42a8209fdab8b36f2f2863509992476296eb3670d0ae3ed7c593858fcf
```
Then I step into the container:

```bash
(local) root@192.168.0.28 ~
$ docker exec -it nagios /bin/sh
/opt/nagios # /opt/nagios/libexec/check_mysql_query -H 10.10.10.100 -u qeep -p secret -q "select 1"
QUERY CRITICAL: Can't connect to MySQL server on '10.10.10.100' (115)
```

So as you can see, the plugin is running normally and no segmantation fault is thrown.

Thanks,
Christos

**Comment by qeepcologne** [comment]
segfault occurs only when database exists and password is correct

**Comment by manios** [comment]
Hi @qeepcologne !

Thanks for the debugging! Indeed it seems that it has something to do with the [musl libc](https://musl.libc.org/) version which causes the issue. Thanks for the debugging!

In my latest image also when I execute:

```shell
/opt/nagios #  ./libexec/check_mysql -P 3306 -H 172.17.0.1 -u root -p root
Uptime: 349  Threads: 1  Questions: 9  Slow queries: 0  Opens: 17  Open tables: 10  Queries per second avg: 0.025|Connections=8c;;; Open_files=22;;; Open_tables=10;;; Qcache_free_memory=1031304;;; Qcache_hits=0c;;; Qcache_inserts=0c;;; Qcache_lowmem_prunes=0c;;; Qcache_not_cached=0c;;; Qcache_queries_in_cache=0;;; Queries=9c;;; Questions=9c;;; Table_locks_waited=0c;;; Threads_connected=1;;; Threads_running=1;;; Uptime=349c;;;
```
against an existing database it works but with `check_mysql_query`, it crashes with Segmentation fault:

```shell
/opt/nagios # ./libexec/check_mysql_query -P 3306 -H 172.17.0.1 -u root -p root -q 'select 1'
Segmentation fault (core dumped)
```
The reason behind that is that the musl version is:
```shell
/opt/nagios # apk info musl
musl-1.2.2-r0 description:
the musl c library (libc) implementation
```
The solution at the moment for us is to release an image which uses `alpine:3.11` as base image which contains `musl-1.1.24-r3` until we have a better answer in Alpine [aports #12601](https://gitlab.alpinelinux.org/alpine/aports/-/issues/12601).

I am preparing a solution with a new build and I will let you know.

Best regards,
Christos

**Comment by manios** [comment]
Hi @qeepcologne !

I have just released a new tag based on `alpine:3.11`. I have tested the `check_mysql_query` plugin with MySQL public servers from [here](). Can you please give it a try

```shell
docker pull manios/nagios:latest
# or
docker pull manios/nagios:4.4.6
```

and provide me with feedback so we can close the issue?

Thanks,
Christos

**Comment by qeepcologne** [comment]
thanks a lot, check_mysql_query is workig fine now.
